db/src/tx.rs

```diff
@@ -749,28 +757,115 @@ fn conclude_tx(tx: Tx, report: TxReport) -> Result<(TxReport, PartitionMap, Opti
     where I: IntoIterator<Item=TermWithTempIds> {
-
     let mut tx = start_tx(conn, partition_map, schema_for_mutation, schema)?;
-    let report = tx.transact_simple_terms(terms, tempid_set)?;
-    conclude_tx(tx, report)
+    let (report, changes) = tx.transact_simple_terms(terms, tempid_set)?;
+    conclude_tx(tx, report, changes)
+}
+
+#[derive(Default, Debug, Clone)]
+pub struct BatchedTransaction {
+    transactions: BTreeMap<Entid, AttributeSet>
+}
+
+impl BatchedTransaction {
+    pub fn add_transact(&mut self, txid: Entid, changes: AttributeSet ) {
+        self.transactions.insert(txid, changes);
+    }
+
+    pub fn get(&self) -> &BTreeMap<Entid, AttributeSet> {
+        &self.transactions
+    }
+}
+
+pub struct TxObserver {
+    notify_fn: Option<Box<FnMut(String, BatchedTransaction)>>
+}
+
+impl TxObserver {
+    pub fn new<F>(notify_fn: F) -> TxObserver where F: FnMut(String, BatchedTransaction) + 'static {
+        TxObserver { notify_fn: Some(Box::new(notify_fn)) }
+    }
+
+    fn notify(&mut self, key: String, transactions: BatchedTransaction) {
+        if let Some(ref mut notify_fn) = self.notify_fn {
+            (notify_fn)(key, transactions);
+        } else {
+            eprintln!("no notify function specified for TxObserver");
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct TxObservationService {
+    observers: BTreeMap<String, TxObserver>,
+    attributes: BTreeMap<String, AttributeSet>,
+}
+
+impl<'o> TxObservationService {
+    // For testing purposes
+    pub fn is_registered(&self, key: &String) -> bool {
+        self.observers.contains_key(key)
+    }
+
+    pub fn register(&mut self, observer: TxObserver, key: String, attributes: AttributeSet) {
+        self.observers.insert(key.clone(), observer);
+        self.attributes.insert(key, attributes);
+    }
+
+    pub fn deregister(&mut self, key: &String) {
+        self.observers.remove(key);
+        self.attributes.remove(key);
+    }
+
+    fn observers_for_attributes(&self, changes: &AttributeSet) -> BTreeSet<(String, AttributeSet)> {
+        // filter observers by ones that care about attributes
+        // add the changes and transaction to the batched transactions of the observers that care about it.
+        self.attributes.iter().filter_map(|(key, attrs)| {
+            let intersect: AttributeSet = attrs.intersection(changes).cloned().collect();
+            if intersect.is_empty() {
+                None
+            } else {
+                Some((key.clone(), intersect))
+            }
+        }).collect()
+    }
+
+    pub fn transaction_did_commit(&mut self, batched_transactions: &Option<BatchedTransaction>) {
+        // notify all observers about their relevant transactions
+        let mut transactions = BTreeMap::new();
+        if let &Some(ref batch) = batched_transactions{
+            for (txid, changes) in batch.get().iter() {
+                for &(ref key, ref attrs) in self.observers_for_attributes(changes).iter() {
```

Comment:
If you simplify `observers_for_attributes`, you can avoid allocating there at all. Just walk the observers and call them.

```
for observer in self.observers.values() {
    if observer.cares_about(batch) {
        observer.tx_function(batch)
    }
}
```